File: modules/signal-protocol/android/src/main/java/expo/modules/signalprotocol/stores/PersistentIdentityKeyStore.kt
package expo.modules.signalprotocol.stores

import android.content.Context
import org.signal.libsignal.protocol.IdentityKey
import org.signal.libsignal.protocol.IdentityKeyPair
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.ecc.ECPublicKey
import org.signal.libsignal.protocol.fingerprint.NumericFingerprintGenerator
import org.signal.libsignal.protocol.state.IdentityKeyStore
import java.security.MessageDigest

class PersistentIdentityKeyStore(
    private val identityKeyPair: IdentityKeyPair,
    private val registrationId: UInt,
    private val keyPrefix: String,
    context: Context
) : IdentityKeyStore {

    private val keystore = KeystoreHelper.getInstance(context)

    override fun getIdentityKeyPair(): IdentityKeyPair {
        return identityKeyPair
    }

    override fun getLocalRegistrationId(): Int {
        return registrationId.toInt()
    }

    override fun saveIdentity(address: SignalProtocolAddress, identityKey: IdentityKey): IdentityKeyStore.IdentityChange {
        val key = identityKeyKey(address)
        val existingIdentity = getIdentity(address)

        // SESSION CORRUPTION FIX: Throw error if save fails
        val saved = keystore.save(key, identityKey.serialize())
        if (!saved) {
            throw RuntimeException("Failed to persist identity to EncryptedSharedPreferences")
        }

        // Return REPLACED_EXISTING if identity replaced an existing different one
        if (existingIdentity != null && !existingIdentity.serialize().contentEquals(identityKey.serialize())) {
            return IdentityKeyStore.IdentityChange.REPLACED_EXISTING
        }
        return IdentityKeyStore.IdentityChange.NEW_OR_UNCHANGED
    }

    override fun isTrustedIdentity(address: SignalProtocolAddress, identityKey: IdentityKey, direction: IdentityKeyStore.Direction): Boolean {
        val savedIdentity = getIdentity(address)

        // TOFU (Trust On First Use): If no identity exists, trust it
        if (savedIdentity == null) {
            return true
        }

        // If the identity matches the saved one, trust it
        if (savedIdentity.serialize().contentEquals(identityKey.serialize())) {
            return true
        }

        // Identity key has changed! Block unless manually verified
        // This causes UntrustedIdentity (Error 12) which the JS layer handles via handleIdentityKeyChanged()
        return isManuallyTrusted(address, identityKey)
    }

    override fun getIdentity(address: SignalProtocolAddress): IdentityKey? {
        val key = identityKeyKey(address)
        val data = keystore.load(key) ?: return null
        return IdentityKey(ECPublicKey(data))
    }

    fun setManuallyTrusted(address: SignalProtocolAddress, identityKey: IdentityKey) {
        val key = trustKey(address, identityKey)
        keystore.save(key, "trusted".toByteArray())
    }

    fun isManuallyTrusted(address: SignalProtocolAddress, identityKey: IdentityKey): Boolean {
        val key = trustKey(address, identityKey)
        return keystore.exists(key)
    }

    /// Generate safety number using standard Signal Protocol NumericFingerprintGenerator
    fun generateSafetyNumber(localUserId: String, remoteAddress: SignalProtocolAddress, remoteIdentity: IdentityKey): String {
        val generator = NumericFingerprintGenerator(5200)
        val localIdentity = identityKeyPair.publicKey
        val fingerprint = generator.createFor(
            2,
            localUserId.toByteArray(Charsets.UTF_8),
            localIdentity,
            remoteAddress.name.toByteArray(Charsets.UTF_8),
            remoteIdentity
        )
        return fingerprint.displayableFingerprint.displayText
    }

    private fun identityKeyKey(address: SignalProtocolAddress): String {
        return "identity-$keyPrefix-${address.name}-${address.deviceId}"
    }

    private fun trustKey(address: SignalProtocolAddress, identityKey: IdentityKey): String {
        // M5 FIX: Use deterministic SHA256 instead of contentHashCode()
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(identityKey.serialize())
        val hashString = hash.take(8).joinToString("") { "%02x".format(it) }
        return "trust-$keyPrefix-${address.name}-${address.deviceId}-$hashString"
    }

}

File: modules/signal-protocol/ios/Stores/PersistentIdentityKeyStore.swift
//
//  PersistentIdentityKeyStore.swift
//  Pods
//
//  TilliT Native — Signal Protocol bindings
//
import Foundation
import LibSignalClient
import CryptoKit

class PersistentIdentityKeyStore: IdentityKeyStore {

    private let identityKeyPair: IdentityKeyPair
    private let registrationId: UInt32
    private let keyPrefix: String
    private let keychain = KeychainHelper.shared

    init(identityKeyPair: IdentityKeyPair, registrationId: UInt32, directoryURL: URL) {
        self.identityKeyPair = identityKeyPair
        self.registrationId = registrationId
        // Use the last path component as prefix for keychain keys
        self.keyPrefix = directoryURL.lastPathComponent
    }

    func identityKeyPair(context: StoreContext) throws -> IdentityKeyPair {
        return identityKeyPair
    }

    func localRegistrationId(context: StoreContext) throws -> UInt32 {
        return registrationId
    }

    func saveIdentity(_ identity: IdentityKey, for address: ProtocolAddress, context: StoreContext) throws -> IdentityChange {
        let key = identityKeyKey(for: address)
        let existingIdentity = try? self.identity(for: address, context: context)

        let data = Data(identity.serialize())
        let saved = keychain.save(data: data, for: key)

        // SESSION CORRUPTION FIX: Throw error if save fails
        if !saved {
            throw SignalError.internalError("Failed to persist identity to Keychain")
        }

        if let existing = existingIdentity {
            return existing.serialize() == identity.serialize() ? .newOrUnchanged : .replacedExisting
        }
        return .newOrUnchanged
    }

    func isTrustedIdentity(_ identity: IdentityKey, for address: ProtocolAddress, direction: Direction, context: StoreContext) throws -> Bool {
        let savedIdentity = try? self.identity(for: address, context: context)

        // TOFU (Trust On First Use): If no identity exists, trust it
        guard let savedIdentity = savedIdentity else {
            return true
        }

        // If the identity matches the saved one, trust it
        if savedIdentity.serialize() == identity.serialize() {
            return true
        }

        // For both RECEIVING and SENDING, block unless manually verified
        // This causes UntrustedIdentity (Error 12) which the JS layer handles via handleIdentityKeyChanged()
        return isManuallyTrusted(identity, for: address)
    }

    func identity(for address: ProtocolAddress, context: StoreContext) throws -> IdentityKey? {
        let key = identityKeyKey(for: address)
        guard let data = keychain.load(for: key) else {
            return nil
        }
        return try IdentityKey(publicKey: PublicKey([UInt8](data)))
    }

    /// Mark an identity key as manually trusted (e.g., after out-of-band verification)
    func setManuallyTrusted(_ identity: IdentityKey, for address: ProtocolAddress) throws {
        let key = trustKey(for: address, identity: identity)
        let data = Data("trusted".utf8)
        _ = keychain.save(data: data, for: key)
    }

    /// Check if an identity key was manually verified by the user
    func isManuallyTrusted(_ identity: IdentityKey, for address: ProtocolAddress) -> Bool {
        let key = trustKey(for: address, identity: identity)
        return keychain.exists(for: key)
    }

    /// Generate safety number using standard Signal Protocol NumericFingerprintGenerator
    func generateSafetyNumber(localUserId: String, for remoteAddress: ProtocolAddress, remoteIdentity: IdentityKey) throws -> String {
        let generator = NumericFingerprintGenerator(iterations: 5200)
        let localKey = identityKeyPair.publicKey   // IdentityKeyPair.publicKey → PublicKey
        let remoteKey = remoteIdentity.publicKey   // IdentityKey.publicKey → PublicKey
        let fingerprint = try generator.create(
            version: 2,
            localIdentifier: Data(localUserId.utf8),
            localKey: localKey,
            remoteIdentifier: Data(remoteAddress.name.utf8),
            remoteKey: remoteKey
        )
        return fingerprint.displayable.formatted
    }

    private func identityKeyKey(for address: ProtocolAddress) -> String {
        return "identity-\(keyPrefix)-\(address.name)-\(address.deviceId)"
    }

    private func trustKey(for address: ProtocolAddress, identity: IdentityKey) -> String {
        // M5 FIX: Use deterministic SHA256 instead of randomized hashValue
        // hashValue in Swift is randomized per app execution, causing trust state to be lost on restart
        let serialized = identity.serialize()
        let hash = SHA256.hash(data: Data(serialized))
        // Use first 8 bytes (16 hex characters) for reasonable key length
        let hashString = hash.prefix(8).map { String(format: "%02x", $0) }.joined()
        return "trust-\(keyPrefix)-\(address.name)-\(address.deviceId)-\(hashString)"
    }
}
